### ISSUE TITLE
Discount Code Expiry Date Problem With 2Checkout (With Extending Only)

### ISSUE THREAD

**Issue by karambk** [strangerstudios/paid-memberships-pro#1531]
**Describe the bug**
we’ve tracked the problem and we now understand that a discount code with 2checkout doesn’t catch the correct expiry date
Only with membership extending, however it was recently fixed with new membership changing as described in this pull https://github.com/strangerstudios/paid-memberships-pro/pull/1510

**To Reproduce**
1. Assign Membership 'Test' for a user
2. Try to extend this membership using 2checkout in that user with a coupon code
3. You will pay the discount code's price, but the expiry date would be incorrect (It would be the membership's original one, without the effect of the coupon code's enddate), you will also see missing 'Fees' field in the members list

**Expected behavior**
1. The expiry date is extended correctly based on the coupon code's settings
2. Fees appear in the members list

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only Paid Memberships Pro plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Memberlite](https://wordpress.org/themes/memberlite/).
- [x] I can reproduce this bug consistently using the steps above.

Link to support ticket :
https://www.paidmembershipspro.com/forums/topic/discount-code-expiry-date-problem-with-2checkout-with-extending-only/

**Comment by karambk** [comment]
I noticed function pmpro_inshandler_level is missing, although it is used in file :
services/twocheckout-ins.php

However, this specific problem doesn't seem to be affected by file 
services/twocheckout-ins.php 

Am confused..

**Comment by dparker1005** [comment]
The main issue here appears to be that the 2Checkout INS handler (which is used to update a user's membership level after completing checkout at 2Checkout) is using the `MemberOrder->getMembershipLevel()` function to get the level object that a user should be changed to:
https://github.com/strangerstudios/paid-memberships-pro/blob/fa57534bb46c9577c3f70c2b674855c986eb6732/services/twocheckout-ins.php#L55

The way that this function works is that it:
1. Checks if the user has a previous membership level entry. If so, load that
2. Build an level from the discount code used if a code was used and the user didn't have a previous membership level entry
3. Build a level from the `pmpro_levels` table using default settings
https://github.com/strangerstudios/paid-memberships-pro/blob/fa57534bb46c9577c3f70c2b674855c986eb6732/classes/class.memberorder.php#L444-L494

The issue that is happening here is that, if a user was an existing customer, the level being loaded is the one that they had previously checked out for, not the one that they are currently checking out for with a discount code. There is also another side effect which is that the `pmpro_checkout_level` filter will not be run (checkout variables that add ons would use on that filter wouldn't be loaded anyways), meaning that any add ons/custom code that modify checkout levels will not be considered when building the membership level object to put into the database.

The best solution for this would probably to use `$_SESSION` variable like we do with PayPal Express, but this is a significant development effort and would take time to implement. In the meantime, the `pmpro_inshandler_level` should then be used to fix the membership level object so that the correct enddate is calculated:
https://github.com/strangerstudios/paid-memberships-pro/blob/fa57534bb46c9577c3f70c2b674855c986eb6732/services/twocheckout-ins.php#L272

**Comment by dparker1005** [comment]
Temporary workaround:
https://gist.github.com/dparker1005/634379279e14b6536ee998cf3f4795b3

NOTE: The level in the order is currently being reset by the discount code getter code after the filter is run. This should be changed so that the discount code is retrieved before the filter is run. Making that change in the core PMPro code is necessary for this workaround to work.

**Comment by dparker1005** [comment]
@karambk I have just submitted a PR which should fix this behavior once it is merged: https://github.com/strangerstudios/paid-memberships-pro/pull/1536

Once that update is pushed out, you should no longer need to use the temporary workaround above.

**Comment by karambk** [comment]
@dparker1005 Understood, I replaced the temporary workaround with the new PR and it is working, I'll keep an eye on this pull until it's merged. thank you David.
